src/gf.c

```diff
@@ -1930,6 +1929,54 @@ jl_value_t *jl_gf_invoke(jl_tupletype_t *types0, jl_value_t **args, size_t nargs
     return jl_call_method_internal(mfunc, args, nargs);
 }
 
+JL_DLLEXPORT jl_value_t *jl_get_invoke_lambda(jl_methtable_t *mt,
+                                              jl_typemap_entry_t *entry,
+                                              jl_tupletype_t *tt)
+{
+    if (!jl_is_leaf_type((jl_value_t*)tt))
+        return jl_nothing;
+    jl_method_t *method = entry->func.method;
+    jl_typemap_entry_t *tm = NULL;
+    if (method->invokes.unknown != NULL) {
+        tm = jl_typemap_assoc_by_type(method->invokes, tt, NULL, 0, 1,
+                                      jl_cachearg_offset(mt));
+        if (tm) {
+            return (jl_value_t*)tm->func.linfo;
+        }
+    }
+
+    JL_LOCK(&method->writelock);
+    if (method->invokes.unknown != NULL) {
+        tm = jl_typemap_assoc_by_type(method->invokes, tt, NULL, 0, 1,
+                                      jl_cachearg_offset(mt));
+        if (tm) {
+            jl_method_instance_t *mfunc = tm->func.linfo;
+            JL_UNLOCK(&method->writelock);
+            return (jl_value_t*)mfunc;
+        }
+    }
+    jl_svec_t *tpenv = jl_emptysvec;
+    jl_tupletype_t *sig = NULL;
+    JL_GC_PUSH2(&tpenv, &sig);
+    if (entry->tvars != jl_emptysvec) {
+        jl_value_t *ti =
+            jl_lookup_match((jl_value_t*)tt, (jl_value_t*)entry->sig, &tpenv, entry->tvars);
+        assert(ti != (jl_value_t*)jl_bottom_type);
+        (void)ti;
+    }
+    sig = join_tsig(tt, entry->sig);
+    jl_method_t *func = entry->func.method;
+
+    if (func->invokes.unknown == NULL)
+        func->invokes.unknown = jl_nothing;
+
+    jl_method_instance_t *mfunc = cache_method(mt, &func->invokes, entry->func.value,
```

Comment:
we may not want to insert into the `invokes` cache, unless `tt` was a leaf signature